### PR TITLE
Feature/metrics

### PR DESCRIPTION
**Pull request by ghkim9213**
This PR
- modifies metrics to take external explain function
- removes unnecessary dependencies on explainer from the metrics 
- adds `optimize` function to support optimization task without `AutoExperiment`

### CODE REVIEW

File: pnpxai/evaluator/metrics/pixel_flipping.py [hunk 1]
```diff
@@ -191,6 +213,8 @@ def __init__(
         baseline_fn: Optional[BaselineFunction]=None,
         prob_fn: Optional[Callable[[Tensor], Tensor]]=lambda outputs: outputs.softmax(-1),
         pred_fn: Optional[Callable[[Tensor], Tensor]]=lambda outputs: outputs.argmax(-1),
+        forward_arg_extractor: Optional[ForwardArgumentExtractor]=None,
+        additional_forward_arg_extractor: Optional[ForwardArgumentExtractor]=None,
     ):
         super().__init__(
             model, explainer, channel_dim, n_steps,
```
Comment (enver1323) [inline pnpxai/evaluator/metrics/pixel_flipping.py:220]: The change looks great, it adds a lot of flexibility, but probably, we need to pass `forward_arg_extractor`, and `additional_forward_arg_extractor` to the `super().__init__()`.

File: pnpxai/evaluator/metrics/pixel_flipping.py [hunk 2]
```diff
@@ -258,6 +284,8 @@ def __init__(
         baseline_fn: Optional[BaselineFunction]=None,
         prob_fn: Optional[Callable[[Tensor], Tensor]]=lambda outputs: outputs.softmax(-1),
         pred_fn: Optional[Callable[[Tensor], Tensor]]=lambda outputs: outputs.argmax(-1),
+        forward_arg_extractor: Optional[ForwardArgumentExtractor]=None,
+        additional_forward_arg_extractor: Optional[ForwardArgumentExtractor]=None,
```
Comment (enver1323) [inline pnpxai/evaluator/metrics/pixel_flipping.py:288]: I think we need to pass these to the parent constructor

File: pnpxai/evaluator/metrics/pixel_flipping.py [hunk 3]
```diff
@@ -327,6 +357,8 @@ def __init__(
...
```
Comment (enver1323) [inline pnpxai/evaluator/metrics/pixel_flipping.py:361]: Here too :)

File: pnpxai/evaluator/metrics/sensitivity.py [hunk 4]
```diff
@@ -21,22 +21,20 @@ class Sensitivity(Metric):
...
-        explainer (Optional[Explainer]): The explainer used for evaluation.
+        explainer (Optional[Union[Explainer, Callable]]): The explainer used for evaluation. It can be an instance of ``Explainer`` or any callable returning attributions from inputs and targets.
         n_iter (Optional[int]): The number of iterations for perturbation.
         epsilon (Optional[float]): The magnitude of random uniform noise.
     """
     def __init__(
         self,
         model: Model,
-        explainer: Optional[Explainer] = None,
+        explainer: Optional[Union[Explainer, Callable]] = None,
```
Comment (enver1323) [inline pnpxai/evaluator/metrics/sensitivity.py:31]: This change looks great!

**Review by enver1323 — CHANGES_REQUESTED** [review]
The PR looks great. I just had couple of concerns, which I mentioned in the comments. Other than that everything should be ready for merge

File: pnpxai/evaluator/metrics/pixel_flipping.py (hunk 1 above)
Comment (ghkim9213) [reply pnpxai/evaluator/metrics/pixel_flipping.py:220]: Thanks for your review. I missed that point. I changed it just now. Additionally, I've considered whether adding the extractors to `Metric`, the base. But they are required only for pixel flipping metrics only, until now. So, in the future, I'll make a separate class if we should add more metrics requiring them.

**Pull request merged**